### PR TITLE
ieee1905_transport: use bind() instead of bpf filter

### PR DESCRIPTION
**Pull request by ghost**
The IEEE1905 transport opens a raw socket to capture 1905.1 and LLDP packets. It currently uses a BPF filter to do that. However, on some platforms, a BPF filter causes the acceleration fabric to stop working or to degrade.

A better solution is to use bind() to tie the raw socket to a specific ethernet protocol. This looks more or less like this:
```c
int openPacketSocket(int ifindex, uint16_t eth_type)
{
    int                 s;
    struct sockaddr_ll  socket_address;

    s = socket(AF_PACKET, SOCK_RAW, htons(eth_type));
    if (-1 == s)
    {
        return -1;
    }

    memset(&socket_address, 0, sizeof(socket_address));
    socket_address.sll_family   = AF_PACKET;
    socket_address.sll_ifindex  = ifindex;
    socket_address.sll_protocol = htons(eth_type);

    if (-1 == bind(s, (struct sockaddr*)&socket_address, sizeof(socket_address)))
    {
        close(s);
        return -1;
    }

    return s;
}
```

Note that a socket can be bound only to 1 ethertype. Therefore, we need to use 2 sockets per interface instead of just one.

### CODE REVIEW

**Comment by tomereli** [comment]
[Replace BPF filter in transport](https://trello.com/c/MMvXUWzL/55-replace-bpf-filter-in-transport)

File: framework/transport/ieee1905_transport/ieee1905_transport_network.cpp [hunk 1]
```diff
@@ -137,50 +81,62 @@ void Ieee1905Transport::update_network_interfaces(
...
-    // and send packets with the Ethernet header included in the buffer. Please consider changing
-    // implementation to use SOCK_DGRAM protocol (without L2 header handling)
-
+    //TODO: Fixup code for SOCK_RAW -> SOCK_DGRAM transition
     // open packet raw socket - see man packet(7) https://linux.die.net/man/7/packet
     int sockfd;
-    if ((sockfd = socket(AF_PACKET, SOCK_RAW, htons(ETH_P_ALL))) < 0) {
-        MAPF_ERR("cannot open raw socket \"" << strerror(errno) << "\" (" << errno << ").");
-        return false;
+    if ((sockfd = socket(AF_PACKET, SOCK_DGRAM, htons(ether_type))) < 0) {
```
Comment (arnout) [inline framework/transport/ieee1905_transport/ieee1905_transport_network.cpp]: We still want SOCK_RAW, otherwise the Ethernet header is stripped off and we parse that.

Note that it's not strictly needed (we can get source and destination address through ancillary data) but it's a lot simpler that way.

File: framework/transport/ieee1905_transport/ieee1905_transport_network.cpp [hunk 2]
```diff
@@ -137,50 +81,62 @@ void Ieee1905Transport::update_network_interfaces(
...
     }
 
-    // the interface can be used by other processes
+    // the interface should be instantly reusable, e.g. after a crash
     int optval = 1;
     if (setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval)) < 0) {
-        MAPF_ERR("cannot set socket option SO_REUSEADDR \"" << strerror(errno) << "\" (" << errno
-                                                            << ").");
+        MAPF_ERR("cannot set socket option SO_REUSEADDR (errno: " << errno << " ["
+                                                                  << strerror(errno) << "])");
```
Comment (arnout) [inline framework/transport/ieee1905_transport/ieee1905_transport_network.cpp:102]: This is an unrelated change. Please do such changes in a separate commit. (nitpick)

File: framework/transport/ieee1905_transport/ieee1905_transport_network.cpp [hunk 3]
```diff
@@ -137,50 +81,62 @@ void Ieee1905Transport::update_network_interfaces(
...
+        return -1;
+    }
+    return sockfd;
+}
+
+bool Ieee1905Transport::open_interface_socket(unsigned int if_index)
+{
+    MAPF_DBG("opening raw socket on interface " << if_index << ".");
+
+    if (network_interfaces_[if_index].ieee1905_socket != -1) {
```
Comment (arnout) [inline framework/transport/ieee1905_transport/ieee1905_transport_network.cpp:128]: This should be in two separate conditions.

though actually, the condition is redundant: close(-1) will return an error, but we ignore its errors anyway.

File: framework/transport/ieee1905_transport/ieee1905_transport_network.cpp [hunk 4]
```diff
@@ -137,50 +81,62 @@ void Ieee1905Transport::update_network_interfaces(
...
+        network_interfaces_[if_index].ieee1905_socket = -1;
+        network_interfaces_[if_index].lldp_socket     = -1;
     }
 
-    network_interfaces_[if_index].fd = sockfd;
+    int ieee1905_socket = open_packet_socket(if_index, ETHERTYPE_IEEE1905);
+    int lldp_socket     = open_packet_socket(if_index, ETHERTYPE_LLDP);
 
+    if (ieee1905_socket < 0 || lldp_socket < 0) {
+        // if only the second call failed, the first socket still needs to be close()d
```
Comment (arnout) [inline framework/transport/ieee1905_transport/ieee1905_transport_network.cpp:139]: And vice-versa. Again, since close(-1) is OK, just close both.

File: framework/transport/ieee1905_transport/ieee1905_transport_network.cpp [hunk 5]
```diff
@@ -137,50 +81,62 @@ void Ieee1905Transport::update_network_interfaces(
...
+    if (network_interfaces_[if_index].ieee1905_socket != -1) {
+        close(network_interfaces_[if_index].ieee1905_socket);
+        close(network_interfaces_[if_index].lldp_socket);
+        network_interfaces_[if_index].ieee1905_socket = -1;
+        network_interfaces_[if_index].lldp_socket     = -1;
     }
 
-    network_interfaces_[if_index].fd = sockfd;
+    int ieee1905_socket = open_packet_socket(if_index, ETHERTYPE_IEEE1905);
+    int lldp_socket     = open_packet_socket(if_index, ETHERTYPE_LLDP);
```
Comment (arnout) [inline framework/transport/ieee1905_transport/ieee1905_transport_network.cpp:136]: There is no need for temporary variables, we can write directly to network_interfaces_[if_index].ieee1905_socket.

File: framework/transport/ieee1905_transport/ieee1905_transport_network.cpp [hunk 6]
```diff
@@ -230,18 +174,23 @@ void Ieee1905Transport::handle_interface_status_change(unsigned int if_index, bo
 
     MAPF_DBG("interface " << if_index << " is now " << (is_active ? "active" : "inactive") << ".");
 
-    if (!is_active && network_interfaces_[if_index].fd >= 0) {
-        poller_.Remove(network_interfaces_[if_index].fd);
-        close(network_interfaces_[if_index].fd);
-        network_interfaces_[if_index].fd = -1;
+    if (!is_active && network_interfaces_[if_index].ieee1905_socket >= 0) {
+        poller_.Remove(network_interfaces_[if_index].ieee1905_socket);
+        poller_.Remove(network_interfaces_[if_index].lldp_socket);
```
Comment (arnout) [inline framework/transport/ieee1905_transport/ieee1905_transport_network.cpp:184]: Condition should be duplicated for lldp_socket

File: framework/transport/ieee1905_transport/ieee1905_transport_network.cpp [hunk 7]
```diff
@@ -230,18 +174,23 @@ void Ieee1905Transport::handle_interface_status_change(unsigned int if_index, bo
...
-    if (is_active && network_interfaces_[if_index].fd < 0) {
+    if (is_active && network_interfaces_[if_index].ieee1905_socket < 0) {
         if (!open_interface_socket(if_index)) {
             MAPF_ERR("cannot open network interface " << if_index << ".");
         }
-        if (network_interfaces_[if_index].fd >= 0)
-            poller_.Add(network_interfaces_[if_index].fd);
+        if (network_interfaces_[if_index].ieee1905_socket >= 0) {
+            poller_.Add(network_interfaces_[if_index].ieee1905_socket);
+            poller_.Add(network_interfaces_[if_index].lldp_socket);
```
Comment (arnout) [inline framework/transport/ieee1905_transport/ieee1905_transport_network.cpp:197]: Here as well, condition must be duplicated.

Yes, it's true that when ieee1905_socket opening failed you also set lldp_socket to -1, but checking them separately is more defensive.

File: framework/transport/ieee1905_transport/ieee1905_transport_network.cpp [hunk 8]
```diff
@@ -230,18 +174,23 @@ void Ieee1905Transport::handle_interface_status_change(unsigned int if_index, bo
...
+        poller_.Remove(network_interfaces_[if_index].lldp_socket);
+        close(network_interfaces_[if_index].ieee1905_socket);
+        close(network_interfaces_[if_index].lldp_socket);
+        network_interfaces_[if_index].ieee1905_socket = -1;
+        network_interfaces_[if_index].lldp_socket     = -1;
     }
 
-    if (is_active && network_interfaces_[if_index].fd < 0) {
+    if (is_active && network_interfaces_[if_index].ieee1905_socket < 0) {
         if (!open_interface_socket(if_index)) {
```
Comment (arnout) [inline framework/transport/ieee1905_transport/ieee1905_transport_network.cpp:192]: Since this function now opens two sockets, perhaps it should be renamed.

File: framework/transport/ieee1905_transport/ieee1905_transport_network.cpp [hunk 9]
```diff
@@ -350,7 +299,8 @@ bool Ieee1905Transport::send_packet_to_network_interface(unsigned int if_index,
 
     packet.header = {.iov_base = &eh, .iov_len = sizeof(eh)};
 
-    int fd             = network_interfaces_[if_index].fd;
+    int fd = packet.ether_type == ETHERTYPE_LLDP ? network_interfaces_[if_index].lldp_socket
```
Comment (arnout) [inline framework/transport/ieee1905_transport/ieee1905_transport_network.cpp:307]: Maybe this should be a switch and should error out if it is neither.

File: framework/transport/ieee1905_transport/ieee1905_transport_network.cpp [hunk 10]
```diff
@@ -83,10 +83,14 @@ void Ieee1905Transport::update_network_interfaces(
 
 int Ieee1905Transport::open_packet_socket(unsigned int if_index, uint16_t ether_type)
 {
+    // Note to developer: The current implementation uses AF_PACKET socket with SOCK_RAW protocol
+    // which means we receive
+    // and send packets with the Ethernet header included in the buffer. Please consider changing
+    // implementation to use SOCK_DGRAM protocol (without L2 header handling)
```
Comment (arnout) [inline framework/transport/ieee1905_transport/ieee1905_transport_network.cpp]: I see no reason to switch to SOCK_DGRAM.
